File: ios/Packages/PadelCore/Sources/PadelCore/Insights/LiveCoach.swift
import Foundation

/// Un aviso corto para la muñeca, a mitad de partido.
public struct LiveTip: Equatable, Sendable {
    /// Identifica la regla, para no repetir el mismo aviso dos veces.
    public let key: String
    /// Lo que se lee de reojo entre puntos. Corto a propósito.
    public let text: String

    public init(key: String, text: String) {
        self.key = key
        self.text = text
    }
}

/// El entrenador en vivo: mira los juegos cerrados y, si hay un patrón claro, suelta un
/// aviso de una línea.
///
/// Es el `InsightEngine` con dos diferencias: pasa **durante** el partido, así que solo
/// puede mirar juegos completos (nunca golpeos, que a mitad de juego no dicen nada), y
/// habla poquísimo. Un aviso a mitad de partido interrumpe; si no cambia lo que vas a
/// hacer en el siguiente juego, no se dice.
///
/// Espejo de `LiveCoach.kt`; los tests viven en el core Kotlin.
public enum LiveCoach {

    /// Sin esto una racha de dos juegos ya "sería un patrón", y no lo es.
    public static let minGames = 4

    /// Tres por lado, uno más que el `InsightEngine` de después del partido: un aviso en
    /// vivo interrumpe, y con dos juegos por lado un 50%-0% es todavía una moneda.
    public static let minGamesPerSide = 3

    /// Diferencia de puntos porcentuales que hace que saque y resto sean dos partidos.
    public static let minSideGap = 40

    /// Juegos seguidos perdidos que merecen un aviso.
    public static let losingStreak = 3

    /// Devuelve el aviso que toca ahora, o nil si no hay nada que decir.
    public static func tip(games: [GameRecord], alreadySaid: Set<String>) -> LiveTip? {
        guard games.count >= minGames else { return nil }

        // La racha manda sobre el resto: es lo más urgente y lo más accionable.
        let streak = games.suffix(losingStreak)
        if streak.count == losingStreak, streak.allSatisfy({ $0.winner == .them }) {
            let key = "racha-\(games.count)"
            if !alreadySaid.contains(key) {
                return LiveTip(
                    key: key,
                    text: "\(losingStreak) juegos seguidos. Cambia el patrón: globo y a la red."
                )
            }
        }

        let serving = games.filter { $0.server == .us }
        let returning = games.filter { $0.server == .them }
        if serving.count >= minGamesPerSide, returning.count >= minGamesPerSide {
            let pctServing = pct(serving)
            let pctReturning = pct(returning)
            if pctServing - pctReturning >= minSideGap, !alreadySaid.contains("saque-mejor") {
                return LiveTip(
                    key: "saque-mejor",
                    text: "Aguantas tu saque (\(pctServing)%) pero al resto vas a "
                        + "\(pctReturning)%. Presiona la primera bola."
                )
            }
            if pctReturning - pctServing >= minSideGap, !alreadySaid.contains("resto-mejor") {
                return LiveTip(
                    key: "resto-mejor",
                    text: "Restando ganas el \(pctReturning)% y sacando el \(pctServing)%. "
                        + "Asegura el primer saque y sube."
                )
            }
        }

        return nil
    }

    private static func pct(_ games: [GameRecord]) -> Int {
        Int((Float(games.filter { $0.winner == .us }.count) * 100 / Float(games.count)).rounded())
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/insights/LiveCoach.kt
package com.risingpadel.core.insights

import com.risingpadel.core.model.GameRecord
import com.risingpadel.core.score.Side

/**
 * Un aviso corto para la muñeca, a mitad de partido.
 *
 * @property key identifica la regla, para no repetir el mismo aviso dos veces.
 * @property text lo que se lee de reojo entre puntos. Corto a propósito.
 */
data class LiveTip(val key: String, val text: String)

/**
 * El entrenador en vivo: mira los juegos cerrados y, si hay un patrón claro, suelta un
 * aviso de una línea.
 *
 * Es el `InsightEngine` con dos diferencias: pasa **durante** el partido, así que solo
 * puede mirar juegos completos (nunca golpeos, que a mitad de juego no dicen nada), y
 * habla poquísimo. Un aviso a mitad de partido interrumpe; si no cambia lo que vas a
 * hacer en el siguiente juego, no se dice.
 *
 * Las mismas reglas de evidencia que el resto: por debajo del mínimo, silencio.
 */
object LiveCoach {

    /** Sin esto una racha de dos juegos ya "sería un patrón", y no lo es. */
    const val MIN_GAMES = 4

    /**
     * Tres por lado, uno más que el `InsightEngine` de después del partido: un aviso en
     * vivo interrumpe, y con dos juegos por lado un 50%-0% es todavía una moneda.
     */
    const val MIN_GAMES_PER_SIDE = 3

    /** Diferencia de puntos porcentuales que hace que saque y resto sean dos partidos. */
    const val MIN_SIDE_GAP = 40

    /** Juegos seguidos perdidos que merecen un aviso. */
    const val LOSING_STREAK = 3

    /**
     * Devuelve el aviso que toca ahora, o null si no hay nada que decir.
     *
     * @param games juegos cerrados en orden de juego.
     * @param alreadySaid claves de avisos ya dados en este partido.
     */
    fun tip(games: List<GameRecord>, alreadySaid: Set<String>): LiveTip? {
        if (games.size < MIN_GAMES) return null

        // La racha manda sobre el resto: es lo más urgente y lo más accionable.
        val streak = games.takeLast(LOSING_STREAK)
        if (streak.size == LOSING_STREAK && streak.all { it.winner == Side.THEM }) {
            val key = "racha-${games.size}"
            if (key !in alreadySaid) {
                return LiveTip(key, "$LOSING_STREAK juegos seguidos. Cambia el patrón: globo y a la red.")
            }
        }

        val serving = games.filter { it.server == Side.US }
        val returning = games.filter { it.server == Side.THEM }
        if (serving.size >= MIN_GAMES_PER_SIDE && returning.size >= MIN_GAMES_PER_SIDE) {
            val pctServing = pct(serving)
            val pctReturning = pct(returning)
            if (pctServing - pctReturning >= MIN_SIDE_GAP && "saque-mejor" !in alreadySaid) {
                return LiveTip(
                    "saque-mejor",
                    "Aguantas tu saque ($pctServing%) pero al resto vas a $pctReturning%. Presiona la primera bola.",
                )
            }
            if (pctReturning - pctServing >= MIN_SIDE_GAP && "resto-mejor" !in alreadySaid) {
                return LiveTip(
                    "resto-mejor",
                    "Restando ganas el $pctReturning% y sacando el $pctServing%. Asegura el primer saque y sube.",
                )
            }
        }

        return null
    }

    // Redondeo, no truncado: es el convenio del resto del core (ver InsightEngine) y
    // el Swift tiene que dar exactamente el mismo número.
    private fun pct(games: List<GameRecord>): Int =
        Math.round(games.count { it.winner == Side.US } * 100.0f / games.size)
}
